truffle/src/main/ruby/core/process_mirror.rb

```diff
@@ -283,10 +347,70 @@ def spawn(options, command, arguments)
           Truffle::Process.spawn command, arguments, env_array, options
         end
 
-        def exec(command, args)
-          Truffle.primitive :vm_exec
-          raise PrimitiveFailure,
-            "Rubinius::Mirror::Process::Execute#exec primitive failed"
+        def exec(command, args, env_array)
+          command.each_byte do |b|
+            if b == 0x00
+              raise ArgumentError.new("string contains null byte")
+            end
+          end
+
+          if args.empty?
+            # If the command contains both the binary to run and the arguments, we need to split them apart. We have
+            # two basic cases here: 1) a fully qualified command; and 2) a simple name expected to be found on the PATH.
+            # Both cases require the split. In the event of a fully qualified command, we exec the command directly,
+            # but the signature for exec requires the command and arguments to all be split. In the other case, where
+            # we have a command to search on the PATH, we must split the command apart from the arguments in order to
+            # perform the search (a poor man's version of shell processing). If we can find it on the PATH, then we
+            # run the whole thing through a shell to get proper shell processing.
+
+            split_command, *split_args = command.strip.split(' ')
+
+            if should_search_path?(split_command)
+              resolved_command = resolve_in_path(split_command)
+
+              if resolved_command
+                command, args = '/bin/sh', ['/bin/sh', '-c', command]
+              else
+                raise Errno::ENOENT.new("No such file or directory - #{command}")
+              end
```

Comment:
Do you mean by running `/usr/bin/env`? Otherwise, in one of these cases we're not supposed to run in the shell at all. In the other case, we'd end up with a different error than MRI (and the specs check for that).